controlplane/kubeadm/internal/workload_cluster_etcd.go

```diff
@@ -88,6 +88,31 @@ func (w *Workload) EtcdIsHealthy(ctx context.Context) (HealthCheckResult, error)
 			response[name] = errors.Wrap(err, "failed to list etcd members using etcd client")
 			continue
 		}
+
```

Comment:
I'm not sure if we should perform these operations during a health check, seems it's not the best place and outside this method's responsibilities, wdyt?

Reply:
That is what I thought too but doing this will cause the need to remember a separate function to delete dangling nodes every time etcd health check is called. So pushed a commit that does it that way but I prefer combining these two.

Reply:
This should be safe to reconcile repeatedly, right? It seems to me like something we'd want in the main "reconcile" flow, not tied specifically to scaling operations.

Reply:
It is safe to be placed in reconcile but only needed if the etcd is unhealthy, so do we want to add this relatively slow logic (needs to connect to etcd) in the reconcile if everything is healthy?

Reply:
I think I'm expecting the KCP to be connecting to etcd on every reconcile anyway to check its health. If that proves troublesome, there's certainly things we can do to speed it up (caching a pool of long-lived connections, for instance).

Reply:
@sethp-nr there were some concerns about doing this: https://github.com/kubernetes-sigs/cluster-api/pull/2849. We can incorporate this in a future PR if found necessary.

Reply:
I share those concerns. Unfortunately, the health checks are doing the heavy lifting for sequencing our upgrades right now. I suspect that means mixing in etcd reconciliation will cause us to get stuck or skip steps. For example, as written, we'll get stuck after removing an etcd member because the remove'd static etcd pod will start crashing and the control plane health check will never pass again.

For what it's worth, I think we can get away with talking to etcd fairly early and often: on average, I expect the probability of an etcd interactions on each reconcile to be barely less than one.

Reply:
Oh, nevermind – we don't check the health of the etcd pod, just the api-server and controller-manager. So it won't get stuck as written, I think.